### PR TITLE
Loose Ruby versioning

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+category: ruby
+image:
+  base: '2019/use-loose-ruby-versioning-in-your-gemfile'
+  alt: 'Red numbered dice'
+  credit:  Terry Vlisidis
+  source: "https://unsplash.com/photos/vPQbo1D7Eco"
+
+---
+
+Patch level versions (e.g. 2.6.**3**) of Ruby are compatible with each other and often only include bug fixes and security patches. Significant upgrades to the Ruby language, that include new syntax, features and make bigger changes to the language are released once a year—as a gift for us all—on Christmas Day.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
Patch level versions (e.g. 2.6.**3**) of Ruby are compatible with each other and often only include bug fixes and security patches. Significant upgrades to the Ruby language that include new syntax or features, and make bigger changes to the language, are released once a year—as a gift for us all—on Christmas Day.
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+Many developers ensure their application use a specific version in development by using a “Ruby version manager”, like [rvm](https://rvm.io) or, my preferred choice, [chruby](https://github.com/postmodern/chruby).
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: *uses

Yesss for `chruby`!

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+These version managers look for a [file](https://gist.github.com/fnichol/1912050) in the root directory of each app called `.ruby-version`. The file specifies the correct version of Ruby for this application and then switches the local environment, switching between multiple installed versions of Ruby automatically.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: *and then automatically switches the local environment to the required version (?)

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+You can also [specify your Ruby version in your Gemfile](https://bundler.io/v1.12/gemfile_ruby.html) using bundler, which is more often used in your deployment environment.
+
+
+## Instead of...
+
+...specifying your exact ruby version in two places and potentially generating conflicts.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: Colon at end?

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+This results in an error, as bundler checks the current version of Ruby.
+
+```
+Your Ruby version is 2.6.5, but your Gemfile specified 2.6.3
+```
+
+## Use
+
+...use loose versioning in the `Gemfile`
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: Colon?

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+
+## Why?
+
+This comes down to a preference for ‘just enough’ flexibility in the `Gemfile`.
+
+You can generally upgrade between 2.6.3 and 2.6.5 without expecting any changes to how your programme works.
+
+I’ve found that having a flexible `Gemfile` Ruby has helped when deploying to environments that pre-build specific patches of Ruby and might not be as completely up to date with the latest patch versions of Ruby as soon as they are released.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: I would start the 'Why?' section from this paragraph, getting rid of the first sentence/paragraph, and moving the second one further down.

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+Examples of this behaviour are Github Actions or Netlify.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: It isn't clear which behaviour you're talking about.

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+The flexible constraint in the `Gemfile` prevents you from having to interrupt your deployment or CI process, whilst being able to upgrade your local Ruby versions.
+
+You also only have to update your specific version in one place when you do make a change.
+
+Additionally, with the flexible constraint, your deployment environment will automatically upgrade your Ruby version, on the next deploy, as it provides a more secure patch-level version.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: This sentence could be clearer.

Maybe, "...with the flexible constraint, when there is a more secure patch-level version, your deployment environment will automatically..." ?

**Review by nodunayo — DISMISSED** [review]

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 2 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
Many developers ensure their application uses a specific version in development by using a “Ruby version manager”, like [rvm](https://rvm.io) or, my preferred choice, [chruby](https://github.com/postmodern/chruby).
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 3 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
These version managers look for a [file](https://gist.github.com/fnichol/1912050) in the root directory of each app called `.ruby-version`. The file specifies the correct version of Ruby for this application and then automatically switches the local environment to the required version.
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 4 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
...specifying your exact ruby version in two places and potentially generating conflicts:
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 8 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
Additionally, with the flexible constraint, when there is a more secure patch-level version, your deployment environment will automatically upgrade your Ruby on the next deploy.
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 5 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
...use loose versioning in the `Gemfile`:
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+category: ruby
+image:
+  base: '2019/use-loose-ruby-versioning-in-your-gemfile'
+  alt: 'Red numbered dice'
+  credit:  Terry Vlisidis
+  source: "https://unsplash.com/photos/vPQbo1D7Eco"
+
+---
+
+Patch level versions (e.g. 2.6.**3**) of Ruby are compatible with each other and often only include bug fixes and security patches. Significant upgrades to the Ruby language that include new syntax or features, and make bigger changes to the language, are released once a year—as a gift for us all—on Christmas Day.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: *Patch versions or Patch-level versions

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+This results in an error, as bundler checks the current version of Ruby.
+
+```
+Your Ruby version is 2.6.5, but your Gemfile specified 2.6.3
+```
+
+## Use
+
+...use loose versioning in the `Gemfile`:
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: *...loose versioning (no double use of 'use')

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+```ruby
+# .ruby-version
+2.6.5
+```
+
+
+## Why?
+
+I’ve found that having a flexible `Gemfile` Ruby has helped when deploying to environments that pre-build specific patches of Ruby and might not be as completely up to date with the latest patch versions of Ruby as soon as they are released.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: Remove use of "I've"

*Having a flexible `Gemfile` Ruby helps when deploying to environments...

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+Examples of services this have a delayed approach to the availability of new Ruby versions include Github Actions and Netlify.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: *that have..

*services where the availability of new Ruby versions is delayed include...

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+You can generally upgrade between 2.6.3 and 2.6.5 without expecting any changes to how your programme works.
```
Comment (nodunayo) [inline _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: *You can usually upgrade between patch-level versions without...

Comment (nodunayo) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]: Or just 'patch versions'...

**Review by nodunayo — DISMISSED** [review]

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 9 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
Patch-level versions (e.g. 2.6.**3**) of Ruby are compatible with each other and often only include bug fixes and security patches. Significant upgrades to the Ruby language that include new syntax or features, and make bigger changes to the language, are released once a year—as a gift for us all—on Christmas Day.
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 10 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
...loose versioning in the `Gemfile`:
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 11 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
Having a flexible `Gemfile` Ruby helps when deploying to environments that pre-build specific patches of Ruby and might not be as completely up to date with the latest patch versions of Ruby as soon as they are released.
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 12 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
Examples of services where the availability of new Ruby versions are delayed include Github Actions and Netlify.
```

File: _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md (hunk 13 above)
Comment (andycroll) [reply _posts/2019/2019-12-15-use-loose-ruby-versioning-in-your-gemfile.md]:
```suggestion
You can generally upgrade between patch-level versions (e.g. 2.6.3 to 2.6.5) without expecting any changes to how your programme works.
```

**Pull request merged**